vendor/riscv/riscv-isa-sim/riscv/Proc.cc

```diff
@@ -92,6 +217,144 @@ st_rvfi Processor::step(size_t n, st_rvfi reference) {
   return rvfi;
 }
 
+void Processor::revert_step(uint32_t num_steps) {
+  FILE *log_file = this->get_log_file();
+
+
+  if (previous_states.size() < num_steps) {
+    throw std::runtime_error("Cannot revert more states than stored");
+  }
+
+  for(auto state: previous_states) {
+    fprintf(log_file, "pc: %lx | ", std::get<0>(state).pc);
+  }
+  fprintf(log_file, "\n");
+
+
+
+  fprintf(log_file, "revert %d steps from PC: %lx", num_steps, this->state.pc);
+
+  prev_changes_t prev_changes = previous_states[num_steps];
+  this->state = std::get<0>(prev_changes);
+
+  fprintf(log_file, " to PC: %lx\n", this->state.pc);
+
+  for (uint32_t i = 0; i <= num_steps; i++) {
+    prev_changes_t prev_changes = previous_states.front();
+    previous_states.pop_front();
+
+    commit_log_mem_t log_mem_pre_write = std::get<1>(prev_changes);
+    fprintf(log_file, "revert mem pc: %lx num: %ld\n", std::get<0>(prev_changes).pc, log_mem_pre_write.size());
+
+    for (auto mem_write : log_mem_pre_write) {
+      fprintf(log_file, "revert mem: addr: %lx val: %lx size: %x", std::get<0>(mem_write), std::get<1>(mem_write), std::get<2>(mem_write));
+      switch (std::get<2>(mem_write))
+      {
+      case 1:
+        this->get_mmu()->store<uint8_t>(std::get<0>(mem_write), (uint8_t)std::get<1>(mem_write),0);
+        break;
+      case 2:
+        this->get_mmu()->store<uint16_t>(std::get<0>(mem_write), (uint16_t)std::get<1>(mem_write),0);
+        break;
+      case 4:
+        this->get_mmu()->store<uint32_t>(std::get<0>(mem_write), (uint32_t)std::get<1>(mem_write),0);
+        break;
+      
+      default:
+        break;
+      }
+      fprintf(log_file, " OK\n");
+    }
+  }
+
+  //Clear commit logs since they contain information from the reverted steps
+  this->get_state()->log_mem_write.clear();
+  this->get_state()->log_reg_write.clear();
+  this->get_state()->log_mem_read.clear();
+  this->get_state()->log_mem_pre_write.clear();
+}
+
+bool Processor::will_trigger_interrupt(reg_t mip) {
+  state_t *state = this->get_state();
+
+  uint32_t old_mip = state->mip->read();
+  uint32_t mie = state->mie->read();
+  uint32_t mstatus = state->mstatus->read();
+  uint32_t old_en_irq = old_mip & mie;
+  uint32_t new_en_irq = mip & mie;
+
+
+  // Only take interrupt if interrupt is enabled, not in debug mode, does not have a halt request, 
+  // and the interrupt is new and not zero
+  if( get_field(mstatus, MSTATUS_MIE) &&
+      !state->debug_mode  &&
+      (this->halt_request != processor_t::HR_REGULAR) &&
+      //(old_en_irq == 0 ) && 
+      (new_en_irq != 0)) 
+  {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool Processor::interrupt(reg_t mip, reg_t mie, uint32_t revert_steps, bool interrupt_allowed) {
+  state_t *state = this->get_state();
+
+  reg_t mask = 0xFFFF0888; // TODO: automatically generate this
```

Comment:
Just flagging all new TODOs in review comments, for visibility.

Reply:
This might not be needed since a similar mask is generated when `mie` is reset